File: code-python/solutions/02_read_solutions.py
# # Reading and Writing Data

# Copyright © 2010–2017 Cloudera. All rights reserved.
# Not to be reproduced or shared without prior written 
# consent from Cloudera.

# This script contains sample solutions to the exercises.  You can run this
# script in a new session or execute selected statements in an existing
# session.  Note that there is generally more than one way to solve a problem.


# ## Setup

# Create a SparkSession:
from pyspark.sql import SparkSession
spark = SparkSession \
  .builder \
  .master("local") \
  .appName("read-solutions") \
  .getOrCreate()


# ## Exercises

# (1) Read the raw driver file from HDFS into a Spark DataFrame.

drivers = spark.read.csv("/duocar/raw/drivers/", header=True, inferSchema=True)

drivers.first()

# (2) Save the driver DataFrame as a JSON file in your CDSW practice directory.

!hdfs dfs -mkdir practice

drivers.write.format("json").save("practice/drivers_json")

# (3) Use Hue to inspect the JSON file.

# **Note:** The JSON file is noticeably larger than the corresponding CSV file.
# Each row is a JSON object.

# (4) Read the driver JSON file into a Spark DataFrame.

drivers_from_json = spark \
  .read \
  .format("json") \
  .load("practice/drivers_json")
  
drivers_from_json.first()

# **Note:** The columns are in alphabetical order.

# (5) Delete the JSON file.
  
!hdfs dfs -rm -r -skipTrash practice/drivers_json


# ## Cleanup

# Stop the SparkSession:

spark.stop()

File: code-r/solutions/02_read_solutions.R
# # Reading and Writing Data


# Sample answers.  You can run this standalone, or execute selected statements 
# in a wider session.


# ## Setup

library(sparklyr)

# Connect local session
config <- spark_config()
config$spark.driver.host <- Sys.getenv("CDSW_IP_ADDRESS")
spark <- spark_connect(
  master = "local",
  app_name = "read-solutions",
  config = config
)

# Reset practice directory in HDFS
system("hdfs dfs -rm -r -skipTrash practice")
system("hdfs dfs -mkdir practice")


# ## Exercises

# Read the raw drivers file into a Spark DataFrame.

# View a few records to verify the file format (could also be done in Hue)
system("hdfs dfs -cat /duocar/raw/drivers/* | head -3")

# Comma-delimited, header present.  Read the csv file.
drivers <- spark_read_csv(
  sc = spark,
  name = "drivers",
  path = "/duocar/raw/drivers/",
  header = TRUE,
  infer_schema = TRUE,
  delimiter = ",")

# Verify
drivers


# Save the drivers DataFrame as a JSON file in your CDSW practice directory.

# (guessing a syntax similar to the one for parquet)
spark_write_json(
  drivers,
  path = "practice/drivers_json")

# Verify (could also be viewed in Hue)
system("hdfs dfs -cat practice/drivers_json/* | head -3")


# Read the drivers JSON file into a Spark DataFrame.

# (guessing syntax similar to the one for parquet)
drivers_json <- spark_read_json(
  sc = spark,
  name = "drivers_json",
  path = "practice/drivers_json")


# Delete the JSON file.

# Cleanup directory from HDFS
system("hdfs dfs -rm -r -skipTrash practice/drivers_json")


# ## Cleanup

# Stop the `SparkSession`
spark_disconnect(spark)
